<!DOCTYPE html>

<head>
    <meta charset="utf-8">
    <style type="text/css">
        * {
            box-sizing: border-box;
            box-shadow: #333;
        }

        h1 {
            margin-bottom: 5px;
            font-size: 1em;
            text-align: center;
            font-weight: 700;
            color: rgb(88, 88, 88);
            padding: 5px 0 0 0;
        }

        .container-width {
            width: 100%;
            margin: 0 auto;
        }

        .flex-sect {
            background-color: #f3f3f3;
            padding: 5px 20px;
            font-family: Helvetica, serif;
        }

        .flex-title {
            margin-bottom: 25px;
            font-size: 2em;
            text-align: center;
            font-weight: 700;
            color: rgb(255, 255, 255);
            padding: 0 5px 0 0;
        }

        .cards {
            padding: 20px 0;
            display: flex;
            justify-content: space-around;
            flex-flow: wrap;
        }

        .form {
            border-radius: 3px;
            padding: 10px 15px;
            background-color: rgba(0, 0, 0, 0.2);
            left: auto;
            justify-content: flex-start;
            flex-direction: row;
            align-items: center;
        }

        .input {
            width: 100%;
            margin-bottom: 15px;
            padding: 7px 10px;
            border-radius: 2px;
            color: #fff;
            background-color: #554c57;
            border: none;
        }

        .label {
            width: 100%;
            display: block;
        }

        .button {
            width: 100%;
            margin: 15px 0;
            background-color: #785580;
            border: none;
            color: #fff;
            border-radius: 2px;
            padding: 7px 10px;
            font-size: 1em;
            cursor: pointer;
        }

        #card_css {
            display: flex;
            position: static;
            right: auto;
        }

        #search_term_form {
            background-image: linear-gradient(#d7d7d7, #d7d7d7);
            background-repeat: repeat;
            background-position: left top;
            background-attachment: scroll;
            background-size: auto;
        }

        #search_term_submit {
            background-image: linear-gradient(#e86464, #e86464);
            background-repeat: repeat;
            background-position: left top;
            background-attachment: scroll;
            background-size: auto;
        }

        #search_term_submit:hover {
            background-image: linear-gradient(#f15555, #f15555);
            background-repeat: repeat;
            background-position: left top;
            background-attachment: scroll;
            background-size: auto;
        }

        #search_term_submit:active {
            background-image: linear-gradient(#38c439, #38c439);
            background-repeat: repeat;
            background-position: left top;
            background-attachment: scroll;
            background-size: auto;
        }

        #search_term_label:active {
            padding: 2px 0 2px 0;
        }

        .navbar-items-c {
            display: inline-block;
            float: left;
        }

        .navbar {
            background-color: rgb(0, 0, 0);
            color: #ddd;
            min-height: 50px;
            width: 100%;
        }

        .navbar-container {
            margin: 0 auto;
            width: 95%;
        }

        .navbar-container::after {
            content: "";
            clear: both;
            display: block;
        }

        .navbar-menu {
            padding: 10px 0;
            display: block;
            float: right;
            margin: 0;
        }

        .navbar-menu-link {
            margin: 0;
            color: inherit;
            text-decoration: none;
            display: inline-block;
            padding: 10px 5px;
        }

        table {
            font-family: arial, sans-serif;
            border-collapse: collapse;
            width: 100%;
        }

        td,
        th {
            border: 1px solid #4c4c4c;
            text-align: left;
            padding: 8px;
        }

        tr:nth-child(even) {
            background-color: #d2d2d2;
        }
    </style>
</head>

<body>
    <div data-gjs="navbar" class="navbar">
        <div class="navbar-container">
            <div id="i12fi" class="navbar-burger">
                <div class="navbar-burger-line">
                </div>
                <div class="navbar-burger-line">
                </div>
                <div class="navbar-burger-line">
                </div>
            </div>
            <div data-gjs="navbar-items" class="navbar-items-c">
                <nav data-gjs="navbar-menu" class="navbar-menu" id="i2rve">
                    <a href="/" class="navbar-menu-link" id="iucgr">Validator Home</a>
                </nav>
            </div>
        </div>
    </div>
    <section class="flex-sect">
        <div class="container-width">
            <div class="flex-title">
                <h1>Validator Findings: 32 issues to review</h1>
            </div>
            <div>
                <table id="results" class="tablesorter">
                    <thead>
                        <tr>
                            <th>
                                Line #
                            </th>
                            <th class="emphasize filter-exact">
                                Error
                            </th><th>
                                Error Source
                            </th>
                            <th>
                                Description
                            </th>
                            <th>
                                Line Content
                            </th>
                        </tr>
                    </thead>
                    <tbody>
                        
                        <tr>
                            <td>3</td>
                            <td>Error</td>
                            <td>'type' (mRNAt) is not valid.</td>
                            <td>This is constrained to be either a term from the Sequence Ontology or an SO accession number.<br><br>To verify the SO term, please use a child_of 'sequence_feature' which can be found here: <a href='http://sequenceontology.org/browser/current_release/term/SO:0000110'>SequenceOntology</a></td>
                            <td>ctg123	.	mRNAt	1299%099000	.	.	.	ID=mrna0001;Name=sonichedgehog;Is_circular=yes;Dbxref="AntWeb:AA816246"</td>
                        </tr>
                        
                        <tr>
                            <td>3</td>
                            <td>Error</td>
                            <td>'start' (1299) is before start bound in ##sequence-region (1300).</td>
                            <td>The start column value must be greater than or equal to the start bound value.</td>
                            <td>ctg123	.	mRNAt	1299%099000	.	.	.	ID=mrna0001;Name=sonichedgehog;Is_circular=yes;Dbxref="AntWeb:AA816246"</td>
                        </tr>
                        
                        <tr>
                            <td>3</td>
                            <td>Error</td>
                            <td>'end' (9000) is after end bound in ##sequence-region (7600).</td>
                            <td>The end column value must be less than or equal to the end bound value.</td>
                            <td>ctg123	.	mRNAt	1299%099000	.	.	.	ID=mrna0001;Name=sonichedgehog;Is_circular=yes;Dbxref="AntWeb:AA816246"</td>
                        </tr>
                        
                        <tr>
                            <td>3</td>
                            <td>Error</td>
                            <td>'attribute 3' (Is_circular=yes) has a non-boolean value for 'Is_circular'.</td>
                            <td>Allowed values are 'true' or 'false' if used.</td>
                            <td>ctg123	.	mRNAt	1299%099000	.	.	.	ID=mrna0001;Name=sonichedgehog;Is_circular=yes;Dbxref="AntWeb:AA816246"</td>
                        </tr>
                        
                        <tr>
                            <td>4</td>
                            <td>Error</td>
                            <td>'phase' (2) is not valid.</td>
                            <td>This feature does not allow for a phase.</td>
                            <td>ctg123	.	exon	1300%091500	.	+	2	ID=exon00001=test;Parent=mrna0001;Dbxref="test_DB:AA816246"</td>
                        </tr>
                        
                        <tr>
                            <td>4</td>
                            <td>Error</td>
                            <td>'attribute 1' (ID=exon00001=test) contains unescaped '=' characters.</td>
                            <td>URL escaping rules are used for tags or values containing the equal sign.</td>
                            <td>ctg123	.	exon	1300%091500	.	+	2	ID=exon00001=test;Parent=mrna0001;Dbxref="test_DB:AA816246"</td>
                        </tr>
                        
                        <tr>
                            <td>4</td>
                            <td>Warning</td>
                            <td>'attribute 3' (test_DB) has an invalid Dbxref value.</td>
                            <td>The list of Dbxref used for this validator can be found at <a href='https://www.ncbi.nlm.nih.gov/genbank/collab/db_xref/'>db_xref</a>.</td>
                            <td>ctg123	.	exon	1300%091500	.	+	2	ID=exon00001=test;Parent=mrna0001;Dbxref="test_DB:AA816246"</td>
                        </tr>
                        
                        <tr>
                            <td>5</td>
                            <td>Error</td>
                            <td>'type' (exon8) is not valid.</td>
                            <td>This is constrained to be either a term from the Sequence Ontology or an SO accession number.<br><br>To verify the SO term, please use a child_of 'sequence_feature' which can be found here: <a href='http://sequenceontology.org/browser/current_release/term/SO:0000110'>SequenceOntology</a></td>
                            <td>ctg123	.	exon8	1650%091500	.	test	.	ID=exon00002;Parent=mrna0001;Ontology_term="GO:0046703"</td>
                        </tr>
                        
                        <tr>
                            <td>5</td>
                            <td>Error</td>
                            <td>'start' (1650) is after 'end' (1500).</td>
                            <td>The start column value must be less than or equal to the end column value.</td>
                            <td>ctg123	.	exon8	1650%091500	.	test	.	ID=exon00002;Parent=mrna0001;Ontology_term="GO:0046703"</td>
                        </tr>
                        
                        <tr>
                            <td>5</td>
                            <td>Error</td>
                            <td>'strand' (test) is not valid.</td>
                            <td>The strand of the feature can be + for positive strand (relative to the landmark), - for minus strand, and . for features that are not stranded. ? can be used for features whose strandedness is relevant, but unknown.</td>
                            <td>ctg123	.	exon8	1650%091500	.	test	.	ID=exon00002;Parent=mrna0001;Ontology_term="GO:0046703"</td>
                        </tr>
                        
                        <tr>
                            <td>6</td>
                            <td>Error</td>
                            <td>'end' (3902b) is not valid.</td>
                            <td>The end coordinate must be a positive 1-based integer.</td>
                            <td>ctg123	.	SO:0000110	3000	3902b	.	+	.	ID=exon00003;Parent=mrna0001;Ontology_term="GO:003b"</td>
                        </tr>
                        
                        <tr>
                            <td>6</td>
                            <td>Warning</td>
                            <td>'end' (3902b) is not valid.</td>
                            <td>The comparison to end cannot be performed.</td>
                            <td>ctg123	.	SO:0000110	3000	3902b	.	+	.	ID=exon00003;Parent=mrna0001;Ontology_term="GO:003b"</td>
                        </tr>
                        
                        <tr>
                            <td>6</td>
                            <td>Error</td>
                            <td>'attribute 3' (GO:003b) is not a valid Ontology_term.</td>
                            <td>Use the GO Browser here to verify the term used: <a href='http://geneontology.org/'>GeneOntology</a></td>
                            <td>ctg123	.	SO:0000110	3000	3902b	.	+	.	ID=exon00003;Parent=mrna0001;Ontology_term="GO:003b"</td>
                        </tr>
                        
                        <tr>
                            <td>7</td>
                            <td>Error</td>
                            <td>'seqid' (ctg 123) contains invalid characters.</td>
                            <td>IDs may not contain unescaped whitespace and must not begin with an unescaped '>'.</td>
                            <td>ctg 123	.	exon	5000a	5500	3.4	-	.	ID=exon00004;Parent=mrna0001</td>
                        </tr>
                        
                        <tr>
                            <td>7</td>
                            <td>Error</td>
                            <td>'start' (5000a) is not valid.</td>
                            <td>The start coordinate must be a positive 1-based integer.</td>
                            <td>ctg 123	.	exon	5000a	5500	3.4	-	.	ID=exon00004;Parent=mrna0001</td>
                        </tr>
                        
                        <tr>
                            <td>7</td>
                            <td>Warning</td>
                            <td>'start' (5000a) is not valid.</td>
                            <td>The comparison to end cannot be performed.</td>
                            <td>ctg 123	.	exon	5000a	5500	3.4	-	.	ID=exon00004;Parent=mrna0001</td>
                        </tr>
                        
                        <tr>
                            <td>8</td>
                            <td>Error</td>
                            <td>'seqid' (>ctg123) contains invalid characters.</td>
                            <td>IDs may not contain unescaped whitespace and must not begin with an unescaped '>'.</td>
                            <td>>ctg123	.	exon	7000	9000	a	?	.	ID=exon00005;Parent=mrna0001</td>
                        </tr>
                        
                        <tr>
                            <td>8</td>
                            <td>Error</td>
                            <td>'end' (9000) is after end bound in ##sequence-region (7600).</td>
                            <td>The end column value must be less than or equal to the end bound value.</td>
                            <td>>ctg123	.	exon	7000	9000	a	?	.	ID=exon00005;Parent=mrna0001</td>
                        </tr>
                        
                        <tr>
                            <td>8</td>
                            <td>Error</td>
                            <td>'score' (a) is not valid.</td>
                            <td>The score of the feature must be a floating point number.</td>
                            <td>>ctg123	.	exon	7000	9000	a	?	.	ID=exon00005;Parent=mrna0001</td>
                        </tr>
                        
                        <tr>
                            <td>9</td>
                            <td>Error</td>
                            <td>'start' (1201) is before start bound in ##sequence-region (1300).</td>
                            <td>The start column value must be greater than or equal to the start bound value.</td>
                            <td>ctg123	.	CDS	1201	1500	.	+	0	ID=cds00001;Parent=mRNA00001=test2</td>
                        </tr>
                        
                        <tr>
                            <td>9</td>
                            <td>Error</td>
                            <td>'attribute 2' (Parent=mRNA00001=test2) contains unescaped '=' characters.</td>
                            <td>URL escaping rules are used for tags or values containing the equal sign.</td>
                            <td>ctg123	.	CDS	1201	1500	.	+	0	ID=cds00001;Parent=mRNA00001=test2</td>
                        </tr>
                        
                        <tr>
                            <td>13</td>
                            <td>Error</td>
                            <td>'start' (1201) is before start bound in ##sequence-region (1300).</td>
                            <td>The start column value must be greater than or equal to the start bound value.</td>
                            <td>ctg123	.	CDS	1201	1500	.	+	0	ID=cds00002;Parent=mRNA00002;Ontology_term="GO:0046703","GO:004b703"</td>
                        </tr>
                        
                        <tr>
                            <td>13</td>
                            <td>Error</td>
                            <td>'attribute 3' (GO:004b703) is not a valid Ontology_term.</td>
                            <td>Use the GO Browser here to verify the term used: <a href='http://geneontology.org/'>GeneOntology</a></td>
                            <td>ctg123	.	CDS	1201	1500	.	+	0	ID=cds00002;Parent=mRNA00002;Ontology_term="GO:0046703","GO:004b703"</td>
                        </tr>
                        
                        <tr>
                            <td>15</td>
                            <td>Error</td>
                            <td>'phase' (.) is not valid.</td>
                            <td>The phase is one of the integers 0, 1, or 2 and is required for all CDS features.</td>
                            <td>ctg123	.	CDS	7000	7600	.	+	.	ID=cds00002;Parent=mRNA00002</td>
                        </tr>
                        
                        <tr>
                            <td>16</td>
                            <td>Error</td>
                            <td>'attribute 3' (Gap=Q8 D3 M6 I1 M6) has an unallowed Gap attribute code.</td>
                            <td>Allowed codes are M, I, D, F, or R.</td>
                            <td>ctg123	.	CDS	3301	3902	.	+	0	ID=cds00003;Parent=mRNA00003;Gap=Q8 D3 M6 I1 M6</td>
                        </tr>
                        
                        <tr>
                            <td>17</td>
                            <td>Warning</td>
                            <td>Line is not tab-delimited for nine columns. Attempting to split columns on other non-tab whitespaces.</td>
                            <td>If columns contained unescaped spaces outside of column 9 (Attributes), the conversion only captures the first 8 whitespaces only. Review the line.</td>
                            <td>ctg123	.	CDS 5000	5500	.	+	1	ID=cds00003;Parent=mRNA00003;Gap=M8 D3 M6 I1 M6</td>
                        </tr>
                        
                        <tr>
                            <td>18</td>
                            <td>Warning</td>
                            <td>'attribute 3' (Target=target id 7000 7005) appears to have unescaped spaces in target_id.</td>
                            <td>If the target_id contains spaces, they must be escaped as hex escape %20.</td>
                            <td>ctg123	.	CDS	7000	7600	.	+	1	ID=cds00003;Parent=mRNA00003;Target=target id 7000 7005</td>
                        </tr>
                        
                        <tr>
                            <td>20</td>
                            <td>Error</td>
                            <td>'attribute 3' (Target=target_id 7000 7005 ?) has a non-allowed target strand.</td>
                            <td>Strand is optional and may be '+' or '-'.</td>
                            <td>ctg123	.	CDS	5000	5500	.	+	1	ID=cds00004;Parent=mRNA00003;Target=target_id 7000 7005 ?</td>
                        </tr>
                        
                        <tr>
                            <td>21</td>
                            <td>Error</td>
                            <td>'end' (7601) is after end bound in ##sequence-region (7600).</td>
                            <td>The end column value must be less than or equal to the end bound value.</td>
                            <td>ctg123	.	CDS	7000	7601	.	+	1	ID=cds00003;Parent=mRNA00003;Target=target_id 7000 6930 +</td>
                        </tr>
                        
                        <tr>
                            <td>21</td>
                            <td>Error</td>
                            <td>'attribute 1' (ID=cds00003) is a non-allowed use of a repeated ID.</td>
                            <td>Note: All lines that share an ID must collectively represent a single feature sequentially.</td>
                            <td>ctg123	.	CDS	7000	7601	.	+	1	ID=cds00003;Parent=mRNA00003;Target=target_id 7000 6930 +</td>
                        </tr>
                        
                        <tr>
                            <td>21</td>
                            <td>Error</td>
                            <td>'attribute 3' (Target=target_id 7000 6930 +) has a target start greater than target end.</td>
                            <td></td>
                            <td>ctg123	.	CDS	7000	7601	.	+	1	ID=cds00003;Parent=mRNA00003;Target=target_id 7000 6930 +</td>
                        </tr>
                        
                        <tr>
                            <td>22</td>
                            <td>Notice</td>
                            <td>All FASTA sequences included in the file must be included together at the end of the file and may not be interspersed with the features lines.</td>
                            <td>Once a ##FASTA section is encountered no other content beyond valid FASTA sequence is allowed.</td>
                            <td>This validator does not check to confirm the FASTA content or if is after the features lines in the GFF3 file. Use caution.</td>
                        </tr>
                        
                    </tbody>
                </table>
            </div>
        </div>
    </section>
</body>

</html>